docs/generated_code.cpp

```diff
@@ -0,0 +1,347 @@
+    [[eosio::action]] void setoracles(std::vector<name>);
+    [[eosio::action]] void resetsuspen(name);
+    [[eosio::action]] void setconfig(asset,uint8_t,uint8_t);
+
+private:
+    /* helpers */
+    config _conf()const{ return config_singleton(get_self(),get_self().value).get(); }
+    void   _ensure_not_paused()const{ check(!_conf().paused_all,"paused"); }
+    void   _ensure_req()const{ const auto& c=_conf(); check(!c.paused_all&&!c.paused_req,"req paused"); }
+    void   _refill(acct_table::iterator,const config&);
+    void   _reward_oracles(asset);
+};
+
+/*--------------------------------------------------*/
+/*  Implementation                                  */
+/*--------------------------------------------------*/
+void orng::_refill(acct_table::iterator it,const config& c){
+    uint32_t maxc=it->stake.amount*c.k_calls_per_wax;
+    uint32_t rate=maxc/3600;
+    uint32_t dt=(current_time_point()-it->last_update).to_seconds();
+    uint32_t add=rate*dt;
+    acct_table at(get_self(),get_self().value);
+    at.modify(it,same_payer,[&](auto&r){
+        r.credits=std::min(r.credits+add,maxc);
+        r.last_update=current_time_point();
+    });
+}
+
+/* stake / unstake / deposit */
+void orng::stake(name d,asset q){
+    _ensure_not_paused(); require_auth(d);
+    check(q.symbol==WAX&&q.amount>0,"bad");
+    acct_table at(get_self(),get_self().value);
+    auto it=at.find(d.value);
+    if(it==at.end()) at.emplace(d,[&](auto&r){r.dapp=d;r.stake=q;r.last_update=current_time_point();});
+    else{ _refill(it,_conf()); at.modify(it,same_payer,[&](auto&r){r.stake+=q;}); }
+}
+void orng::unstake(name d,asset q){
+    _ensure_not_paused(); require_auth(d);
+    acct_table at(get_self(),get_self().value);
+    auto it=at.require_find(d.value,"?"); _refill(it,_conf());
+    check(it->stake>=q,"exceed");
+    at.modify(it,same_payer,[&](auto&r){ r.stake-=q; });
+    action{ {get_self(),"active"_n},TOKEN,"transfer"_n,
+            std::make_tuple(get_self(),d,q,string("unstake")) }.send();
+}
+void orng::deposit(name d,asset q){
+    _ensure_not_paused(); require_auth(d);
+    check(q.symbol==WAX&&q.amount>0,"bad");
+    acct_table at(get_self(),get_self().value);
+    auto it=at.find(d.value);
+    if(it==at.end()) at.emplace(d,[&](auto&r){r.dapp=d;r.fee_balance=q;r.last_update=current_time_point();});
+    else at.modify(it,same_payer,[&](auto&r){r.fee_balance+=q;});
+}
+
+/* requestrand */
+void orng::requestrand(name d,checksum256 seed,uint64_t assoc){
+    _ensure_req(); require_auth(d);
+    ban_table bt(get_self(),get_self().value); check(bt.find(d.value)==bt.end(),"banned");
+
+    config cfg=_conf();
+    acct_table at(get_self(),get_self().value);
+    auto it=at.require_find(d.value,"stake first"); _refill(it,cfg);
+
+    if(it->credits==0){
+        check(it->fee_balance>=cfg.fee_per_call,"need fee");
+        at.modify(it,same_payer,[&](auto&r){ r.fee_balance-=cfg.fee_per_call;});
+        _reward_oracles(cfg.fee_per_call);
```

Comment:
should we be doing this here? It looks like we _reward_oracles at the bottom of the setrand action. I think that should be all we need?It looks to me like we might be double rewarding the oracles when the dapp uses payp per call model with this here.